src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/W3CPropagator.cs

```diff
@@ -0,0 +1,745 @@
+                }
+
+                processed += delta;
+            }
+
+            if (processed > 0)
+            {
+                if (traceState[processed - 1] == Comma)
+                {
+                    processed--; // remove the last comma
+                }
+
+                if (processed > 0)
+                {
+                    return processed >= traceState.Length ? traceState : traceState.AsSpan(0, processed).ToString();
+                }
+            }
+
+            return null;
+        }
+
+        internal static void InjectTraceState(string traceState, object? carrier, PropagatorSetterCallback setter)
+        {
+            Debug.Assert(traceState != null, "traceState cannot be null");
+            Debug.Assert(setter != null, "setter cannot be null");
+
+            string? traceStateValue = ValidateTraceState(traceState);
+            if (traceStateValue is not null)
+            {
+                setter(carrier, TraceState, traceStateValue);
+            }
+        }
+
+        internal static void InjectW3CBaggage(object? carrier, IEnumerable<KeyValuePair<string, string?>> baggage, PropagatorSetterCallback setter)
+        {
+            using (IEnumerator<KeyValuePair<string, string?>> e = baggage.GetEnumerator())
+            {
+                if (e.MoveNext())
+                {
+                    ValueStringBuilder encodedBaggage = new ValueStringBuilder(stackalloc char[256]);
+
+                    int entriesCount = 0;
+                    int lastGoodLength = 0;
+
+                    do
+                    {
+                        KeyValuePair<string, string?> item = e.Current;
+
+                        if (EncodeBaggageKey(item.Key, ref encodedBaggage))
+                        {
+                            encodedBaggage.Append(Space);
+                            encodedBaggage.Append(Equal);
+                            encodedBaggage.Append(Space);
+                            if (!string.IsNullOrEmpty(item.Value))
+                            {
+                                EncodeBaggageValue(item.Value, ref encodedBaggage);
+                            }
+                            encodedBaggage.Append(CommaWithSpace);
+
+                            entriesCount++;
+
+                            if (encodedBaggage.Length < MaxBaggageEncodedLength)
+                            {
+                                lastGoodLength = encodedBaggage.Length;
+                            }
+                        }
+                    } while (e.MoveNext() && entriesCount < MaxBaggageEntriesToEmit && encodedBaggage.Length < MaxBaggageEncodedLength);
+
+                    if (lastGoodLength - 2 > 0)
+                    {
+                        setter(carrier, Baggage, encodedBaggage.AsSpan(0, lastGoodLength - 2).ToString());
+                    }
+                    else
+                    {
+                        encodedBaggage.Dispose();
+                    }
```

Comment:
```suggestion

                    encodedBaggage.Dispose();
```
The `if` block also needs to return the (potentially) pooled buffer since it's using `AsSpan` (only `ToString` returns it for you).